fedimint-client/src/module/mod.rs

```diff
@@ -22,51 +22,117 @@ use fedimint_core::task::{MaybeSend, MaybeSync};
+    PeerId, TransactionId,
 };
-use fedimint_eventlog::Event;
+use fedimint_eventlog::{Event, EventKind};
+use fedimint_logging::LOG_CLIENT;
 use futures::Stream;
 use serde::de::DeserializeOwned;
 use serde::Serialize;
+use tracing::warn;
 
 use self::init::ClientModuleInit;
 use crate::module::recovery::{DynModuleBackup, ModuleBackup};
-use crate::oplog::{OperationLogEntry, UpdateStreamOrOutcome};
-use crate::sm::{self, ActiveStateMeta, Context, DynContext, DynState, State};
+use crate::oplog::{OperationLog, OperationLogEntry, UpdateStreamOrOutcome};
+use crate::sm::{self, ActiveStateMeta, Context, DynContext, DynState, Executor, State};
 use crate::transaction::{ClientInputBundle, ClientOutputBundle, TransactionBuilder};
 use crate::{
-    oplog, AddStateMachinesResult, Client, ClientStrong, ClientWeak,
-    InstancelessDynClientInputBundle, TransactionUpdates,
+    oplog, AddStateMachinesResult, Client, InstancelessDynClientInputBundle, TransactionUpdates,
 };
 
 pub mod init;
 pub mod recovery;
 
 pub type ClientModuleRegistry = ModuleRegistry<DynClientModule>;
 
+/// A fedimint-client interface exposed to client modules
+///
+/// To break the dependency of the client modules on the whole fedimint client
+/// and in particular the `fedimint-client` crate, the module gets access to an
+/// interface, that is implemented by the `Client`.
+///
+/// This allows lose coupling, less recompilation and better control and
+/// understanding of what functionality of the Client the modules get access to.
+#[apply(async_trait_maybe_send!)]
+pub trait ClientContextIface: MaybeSend + MaybeSync {
+    fn get_module(&self, instance: ModuleInstanceId) -> &maybe_add_send_sync!(dyn IClientModule);
+    fn api_clone(&self) -> DynGlobalApi;
+    fn decoders(&self) -> &ModuleDecoderRegistry;
+    async fn finalize_and_submit_transaction(
+        &self,
+        operation_id: OperationId,
+        operation_type: &str,
+        operation_meta_gen: Box<maybe_add_send_sync!(dyn Fn(OutPointRange) -> serde_json::Value)>,
+        tx_builder: TransactionBuilder,
+    ) -> anyhow::Result<OutPointRange>;
+
+    // TODO: unify
+    async fn finalize_and_submit_transaction_inner(
+        &self,
+        dbtx: &mut DatabaseTransaction<'_>,
+        operation_id: OperationId,
+        tx_builder: TransactionBuilder,
+    ) -> anyhow::Result<OutPointRange>;
+
+    async fn transaction_updates(&self, operation_id: OperationId) -> TransactionUpdates;
+
+    async fn await_primary_module_outputs(
+        &self,
+        operation_id: OperationId,
+        // TODO: make `impl Iterator<Item = ...>`
+        outputs: Vec<OutPoint>,
+    ) -> anyhow::Result<()>;
+
+    fn operation_log(&self) -> &OperationLog;
+
+    async fn has_active_states(&self, operation_id: OperationId) -> bool;
+
+    async fn operation_exists(&self, operation_id: OperationId) -> bool;
+
+    async fn config(&self) -> ClientConfig;
+
+    fn db(&self) -> &Database;
+
+    fn executor(&self) -> &Executor;
```

Comment:
We'd have to move the `Executor` around too. Maybe me making it generic over the context will finally pay off :laughing:

Reply:
@elsirion Now that we have an explicit interface, we can iterate about making it sane. :D